### ISSUE TITLE
possible bug in zipkin-ruby or in the visualization

### ISSUE THREAD

**Issue by jcarres-mdsol** [openzipkin/zipkin#2414]
After updating to 2.12.2 we have seen that many of our traces get smashed to the left of the graph
![screen shot 2019-02-25 at 11 41 28 am](https://user-images.githubusercontent.com/1803073/53363826-5a50da00-38f2-11e9-8816-a638ba8337ff.png)

When clicking on one of those I get:

![screen shot 2019-02-25 at 11 42 36 am](https://user-images.githubusercontent.com/1803073/53363908-7eacb680-38f2-11e9-9646-543ae893540b.png)

Only SA information.
Looking at the json, these could be the relevant sections:
```
  {
    "traceId": "7309633b16097a5c",
    "parentId": "7309633b16097a5c",
    "id": "4f829b6d0eb7bfb5",
    "name": "post",
    "localEndpoint": {
      "serviceName": "hermes-sandbox",
      "ipv4": "54.162.48.195",
      "port": 443
    },
    "tags": {
      "sa": "1"
    }
  },
```

```
  {
    "traceId": "7309633b16097a5c",
    "parentId": "7309633b16097a5c",
    "id": "4f829b6d0eb7bfb5",
    "kind": "SERVER",
    "name": "post /path_info_in_here_edited_out",
    "timestamp": 1551123015539229,
    "duration": 999430,
    "localEndpoint": {
      "serviceName": "hermes-sandbox",
      "ipv4": "172.17.0.20",
      "port": 3000
    },
    "tags": {
      "http.path": "/apps/exd/api/mauth_proxy"
    }
  },
```

Note how the SA notation is alone in a different section of the JSON even when they have the same ID.

I assume the UI is rendering any one of the other depending on the orther it finds this information. Thus it gets an annotation with no further info and gets rendered this way.

These traces are being sent by the zipkin-ruby tracer. I suspect a bug there,  but please verify.
Zipkin-ruby does *not* have support for v2 traces yet which can be the root of this.

If you determine this is a bug in the current UI, I do not care really as long as Lens do not have the same bug as we will migrate as soon as it is available in the stable image.

**Comment by jcarres-mdsol** [comment]
CC @johnduhart

**Comment by codefromthecrypt** [comment]
what this smells like is v1 json encoding 1 instead of true for the "sa"
binary annotation.. could that be the case? in thrift boolean true is
encoded as a byte with one bit set..

**Comment by codefromthecrypt** [comment]
"sa" aside it looks like the span is sent in two documents instead of one.
it is odd that the IP address would change during the same request on the
same host. usually that is symptomatic of a loopback RPC. Even if it
weren't, we would have to guess which side of the RPC to stick the extra
span doc (the client or the server side).

thanks for posting.. will think about it. meanwhile if there is a way to
keep stable endpoint info and post a single doc instead of two that will be
good anyway.

**Comment by jcarres-mdsol** [comment]
We are hardcoding `sa` to `1` somehow we thought that was the way. We'll change that to `true` (boolean or the string "true"?) if you can verify.
More weird still, I've looked at the code and I see where we set the `sa`, it is in the client span, which in fact it is in the trace also. At that point in the code there are other annotations, and they do show up:

```
  {
    "traceId": "7309633b16097a5c",
    "parentId": "4f829b6d0eb7bfb5",
    "id": "d1887feaeb2591de",
    "kind": "CLIENT",
    "name": "post",
    "timestamp": 1551123015543929,
    "duration": 993347,
    "localEndpoint": {
      "serviceName": "hermes-sandbox",
      "ipv4": "172.17.0.20",
      "port": 3000
    },
    "tags": {
      "http.method": "POST",
      "http.path": "/v1/protocol_versions/94365fe0-d84d-4ed7-af48-3991b12bde4a/publish",
      "http.status": "200",
      "sa": "1"
    }
  },
```

and nowhere else the sa is setup, so really no idea where this particular sa with no extra info comes from.

So to summarize, we have 3 pieces here , a SERVER, a CLIENT, both I can find in the client code. And a SA which has no extra info which I can't find.
Also we have the SA == 1 thing which if you tell me so I'll change to the boolean true.

**Comment by codefromthecrypt** [comment]
> We are hardcoding sa to 1 somehow we thought that was the way. We'll
change that to true (boolean or the string "true"?) if you can verify.

definitely boolean true (even if string might work)
https://github.com/openzipkin/zipkin-api/blob/master/zipkin-api.yaml#L298

More weird still, I've looked at the code and I see where we set the sa, it
> is in the client span, which in fact it is in the trace also. At that point
> in the code there are other annotations, and they do show up:
>
>   {
>     "traceId": "7309633b16097a5c",
>     "parentId": "4f829b6d0eb7bfb5",
>     "id": "d1887feaeb2591de",
>     "kind": "CLIENT",
>     "name": "post",
>     "timestamp": 1551123015543929,
>     "duration": 993347,
>     "localEndpoint": {
>       "serviceName": "hermes-sandbox",
>       "ipv4": "172.17.0.20",
>       "port": 3000
>     },
>     "tags": {
>       "http.method": "POST",
>       "http.path": "/v1/protocol_versions/94365fe0-d84d-4ed7-af48-3991b12bde4a/publish",
>       "http.status": "200",
>       "sa": "1"
>     }
>   },
>
> and nowhere else the sa is setup, so really no idea where this particular
> sa with no extra info comes from.
>
not sure.. maybe when late route info is sent.

> So to summarize, we have 3 pieces here , a SERVER, a CLIENT, both I can
> find in the client code. And a SA which has no extra info which I can't
> find.
> Also we have the SA == 1 thing which if you tell me so I'll change to the
> boolean true.
>
So, the reason why the "sa" gets separated is  "sa" means "server address"
and that is a client attribute. "sa" should never be set on a server span.
So, when the attribute is sent standalone, the parser assumes this is a
client span, as if a remote address in v1 format was set on a server span,
it would be "ca" (client address) not "sa" (server address)

obv doing native v2 encoding would be  a win at some point (nudge nudge) as
that gets rid of the whole awkwardness and confusion around "sa" etc, as
remoteAddress is harder to get wrong.

**Comment by jcarres-mdsol** [comment]
The only place where I see SA  set is here

```
        span.record_tag(Trace::BinaryAnnotation::METHOD, method.upcase, Trace::BinaryAnnotation::Type::STRING, local_endpoint)
        span.record_tag(Trace::BinaryAnnotation::PATH, url.path, Trace::BinaryAnnotation::Type::STRING, local_endpoint)
        span.record_tag(Trace::BinaryAnnotation::SERVER_ADDRESS, SERVER_ADDRESS_SPECIAL_VALUE, Trace::BinaryAnnotation::Type::BOOL, remote_endpoint)
        span.record(Trace::Annotation::CLIENT_SEND, local_endpoint)
        response = @app.call(env).on_complete do |renv|
          record_response_tags(span, renv[:status].to_s, local_endpoint)
        end
        span.record(Trace::Annotation::CLIENT_RECV, local_endpoint)
```

Which is a client span. I never see it in a server span.
Yeah I want to do V2 also, not enough hands

**Comment by codefromthecrypt** [comment]
ok so in the initial example, it looks like maybe the data was missing some of the client part (only getting the "sa" info).. weird. anyway this is more a troubleshooting type of conversation. We should probably not do this in an issue as it spams everyone watching. Let's move to gitter? https://gitter.im/openzipkin/zipkin

**Comment by jcarres-mdsol** [comment]
It is not missing information, the client span is there, the server span is there, and then there is this extra sa thing which I can't see where is coming from

**Comment by codefromthecrypt** [comment]
Below is the complete json that creates this concern. I was able to reproduce it with the canonical frontend/backend app hacked here https://github.com/openzipkin/zipkin-ruby-example

The only option I can think of in span conversion are hacks:
* for address keyword in "sa" "ca" "ma"
  * leniently treat any binary annotation with address keyword as truthy (ex ignore the value which would let `"1"` be same as `true`. The value is actually unimportant.
  * drop any binary annotation which has address keyword and a value besides true.

@openzipkin/core @drolando any opinons?

```json
[
  {
    "traceId": "23ecf2011083ef43",
    "parentId": "23ecf2011083ef43",
    "id": "e323382e9c9a9653",
    "kind": "SERVER",
    "name": "get",
    "timestamp": 1551151784864992,
    "duration": 5466,
    "localEndpoint": {
      "serviceName": "backend",
      "ipv4": "192.168.0.107",
      "port": 80
    }
  },
  {
    "traceId": "23ecf2011083ef43",
    "id": "23ecf2011083ef43",
    "kind": "SERVER",
    "name": "get",
    "timestamp": 1551151784856589,
    "duration": 21271,
    "localEndpoint": {
      "serviceName": "frontend",
      "ipv4": "192.168.0.107",
      "port": 80
    },
    "tags": {
      "http.path": "/"
    }
  },
  {
    "traceId": "23ecf2011083ef43",
    "parentId": "23ecf2011083ef43",
    "id": "e323382e9c9a9653",
    "kind": "CLIENT",
    "name": "get",
    "timestamp": 1551151784857207,
    "duration": 14067,
    "localEndpoint": {
      "serviceName": "frontend",
      "ipv4": "192.168.0.107",
      "port": 80
    },
    "tags": {
      "http.method": "GET",
      "http.path": "/api",
      "http.status": "200"
    }
  },
  {
    "traceId": "23ecf2011083ef43",
    "parentId": "23ecf2011083ef43",
    "id": "e323382e9c9a9653",
    "name": "get",
    "localEndpoint": {
      "serviceName": "backend",
      "ipv4": "127.0.0.1",
      "port": 9000
    },
    "tags": {
      "sa": "1"
    }
  }
]
```

**Comment by codefromthecrypt** [comment]
fwiw personally I am fine with conversions of address keywords ignoring the binary annotation value. As you can see in the image of the original trace, mistaking remote address for local one really wrecks data conversion.

**Comment by codefromthecrypt** [comment]
https://github.com/openzipkin/zipkin/pull/2415 should work around this, but fixing the encoding should go out asap
